File: lc/coreprep/implementation/patterns_reference_implementation/modified_binary_search/floor.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(logn) / O(1)
int floor(const vector<int> &a, int low, int high, int k) {
  if (a[0] > k) {
    return -1;
  }

  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] < k) {
      low = mid + 1;
    } else if (a[mid] > k) {
      high = mid - 1;
    } else {
      return mid;
    }
  }

  return high;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << floor(a, 0, n - 1, k) << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/Floor.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class Floor {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        System.out.println(floor(a, 0, n - 1, k));
      }
    }
  }

  // O(logn) / O(1)
  public static int floor(int[] a, int low, int high, int k) {
    if (a[low] > k) {
      return -1;
    }

    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        return mid;
      }
    }

    return high;
  }
}